### ISSUE TITLE
Describe API using API Blueprint

### ISSUE THREAD

**Issue by kpande** [gogits/go-gogs-client#19]
Available [here](https://apiblueprint.org), API Blueprint is a specification to describe an API. Swagger and API Blueprint are two such specifications that have a wide variety of tooling available for use with the spec.

For example, Apiary @ apiary.io creates a Mock Server that allows API "testing".

`Hiro` is a Golang static page generator for API Blueprints.

`Dredd` is an automated test suite for API regression / functionality / documentation. It uses the API Blueprint to test the API and ensure it works like the doc says it will.

**Comment by Unknwon** [comment]
Actually I'm thinking http://raml.org/ as well, thoughts?

**Comment by kpande** [comment]
RAML makes it difficult to describe complex APIs (in my experience)

**Comment by Unknwon** [comment]
> RAML makes it difficult to describe complex APIs (in my experience)

How does blueprint solve that in your case?

**Comment by kpande** [comment]
The API Blueprint community has their `MSON` spec, Markdown serialized object notation which makes nesting really easy. If you want a change made, an RFC quickly gets feedback and it's easier to upstream these changes.

I have written one here, which you can see by clicking "View API Blueprint" at top left:
http://docs.panelapi.apiary.io/

It uses API Blueprint Objects to provide reusable model.

There is an API converter [here](https://apitransformer.com/) to change formats but it can lose something on translation. If you find an issue, they are responsive and fix it.

**Comment by Unknwon** [comment]
Looks very neat. My question is does it only support one page?

**Comment by kpande** [comment]
It is up to you. Different tools do different things.

For my usage, I use different independent files for each endpoint and then a single index that is processed by `aglio` into a single `.md`. A tool exists called `Hercule` which uses somewhat different syntax than `aglio`.

You can certainly do your own rendering where each file is generated into its own markup.

Apiary is a hosted provider for rendered files, they only support one page. You can use something like `hiro` or `aglio` to get multiple.

**Comment by Unknwon** [comment]
@kpande thanks!

**Comment by kpande** [comment]
Also, there is (i think golang-based) mock server that generates a fake API service based on your blueprint. For sure it is in node.js (npm)

**Comment by kpande** [comment]
thanks! gogs is wicked! i think these api descriptions (whichever language) are the best way to provide developers with information on your APIs. they can always import the Blueprint in their own tool or convert to different format if needed.

Cheers

**Comment by Unknwon** [comment]
Seems like there is no available tool for multi-page documentation for both API Blueprint and RAML.

**Comment by kpande** [comment]
https://github.com/jamesramsay/hercule

**Comment by kpande** [comment]
https://github.com/peterhellberg/hiro

**Comment by Unknwon** [comment]
For a quick read, I think they only support output to a single HTML file.

**Comment by kpande** [comment]
you will need to do something like this:

file1.md:
```
<!-- include('category1.apib'); -->
<!-- include('category2.apib'); -->
<!-- include('category3.apib'); -->
<!-- include('category4.apib'); -->
```

file2.md:
```
<!-- include('category5.apib'); -->
<!-- include('category6.apib'); -->
<!-- include('category7.apib'); -->
<!-- include('category8.apib'); -->
```


file1 and file2 become indexes of categories.

`aglio -i file1.md -o file1.html` < generate first page
`aglio -i file2.md -o file2.html` < generate second page
`aglio -i file1.md -o file1_index.md -c` < compile multiple files into one md

**Comment by Unknwon** [comment]
Thanks! I'll try that.

**Comment by bibo38** [comment]
I would like to see a way to automatically generate the code from the API Specification.
For example: The [Swagger Editor](http://editor.swagger.io) is capable to generate the client and server code for common languages like Java, Javascript, PHP, Python, Go, C++(with Qt 5).

It would be nice to have this feature availiable to save the time of reimplementing the API in the desired language and keeping up with the changes made to it.
This approach reduces time, synchronisation problems and implementation bugs.

P.S.: Swagger offers the ability to define tags, which you can use to filter the API. Also you can generate a DHTML-Document, which is a Node.js-Application, where each tag has its **own page**.
This looks like this(the tags are estimates, products, user):
![ss](https://cloud.githubusercontent.com/assets/1812304/13901551/feb21604-ee27-11e5-852d-c9e75314574b.png)

**Comment by kpande** [comment]
I disagree. Adding API documentation within the code will lead to messy, hard-to-read code. This is why I suggested API Blueprint instead of Swagger. Both are able to be written independently of code generators which is a preferable method of description. You can't describe all the possible GET and POST etc requests without significantly bloating your code. Have you actually *used* that?

**Comment by bibo38** [comment]
Where did I say, that I want to add the Documentation in the code?
But I would prefer to have a Javadoc generated with the generated code, because any Java IDE would be able to show me the documentation of an API request.

Or do you mean, that I would have unused classes/functions in my project, because I only use the one part of the API? Otherwise I don't really understand the Code bloating.

**Comment by kpande** [comment]
Ah, my apologies - I misread what you typed. I tried using swagger for PHP initially which has the flawed assumption of adding API specification to the project code directly via DocBlocks.

**Comment by kpande** [comment]
I have worked with these guys to help ensure document compatibility across languages. We can convert APIB to RAML and Swagger formats: https://apitransformer.com/

**Comment by Unknwon** [comment]
To be honest, I prefer hand written API docs.

**Comment by wyattjoh** [comment]
I've used API Blueprint quite extensively at my company, but found it pretty restricting and complicated to add and extend types. I've tried out Swagger recently, and am currently using it for an existing project, and really do recommend it.

**Comment by einarf** [comment]
I prefer hand written API docs as well, but I've seen the value in systems like swagger so others can use that to generate complete or partial code for communicating with your API and almost any language.

**Comment by iwanbk** [comment]
@Unknwon 
```
Actually I'm thinking http://raml.org/ as well, thoughts?
```
raml1.0 was released a few months ago, i think you need to check it out again.

@kpande 
```
RAML makes it difficult to describe complex APIs (in my experience)
```
RAML 1.0 has support for libraries, traits, inheritance, resource types etc....you might need to check it out again.

**Comment by BryanDollery** [comment]
I think Swagger has the most users and most support, so if you want to maximise the number of devs who can contribute to the project you should probably choose that.

**Comment by patrickmcclory** [comment]
I can't stress this enough.. it's absolutely a good idea to do this and I'd be more than happy to help. Some major benefits here would include: 

* common, versioned and stable API structure that can be leveraged from any language
  * levels the detail up from code-level implementation to protocol.
* easy client generation for multiple languages
* platform-independent means of articulating and therefore being able to consume the api.
* builds a strong separation of duties and allows for the api-end of things to facade the 'guts' of the process such that clients are shielded from shifting/changing/binary updates if/when non-breaking (non-contract-breaking) changes happen. 

There are a few issues with how things are put together today that make this a concern for me... 

* I'm absolutely confused as to why the hook struct definitions exist in the client library (this one) and they're referenced by the server (gogs main project).  IMHO this is inverted... 
* The api is easily modeled and there's a bunch of detail around the project already that we're aspiring to a Github v3-like setup... this seems like a great goal... and sure, there are going to be things that are variations on the theme, but we should start with the basics and go from there. 
* finally, we need to make sure that we get the contract defined and then that the detail is built into the Gogs project directly. 
* with that, the api may actually be better served as a separate project that inherits from the main gogs project itself so we could do a good bit of codegen (for stubs and models)... 

I'll be honest--I've followed design-by-contract with many teams that I've both been a part of and built/managed and I feel fairly confident that we can get this set up such that it's easy to manage.  I'd be absolutely happy to lead the first effort in defining the contract, but I'm very much more of a python dev aspiring to work more in Golang (which I have some decent experience... but I wouldn't want to be the only one on it :) ).  

If this group is interested, I'll incubate a contract (or set of contract docs) in swagger 2.0 and we can go from there?

**Comment by patrickmcclory** [comment]
To back up a bit .. my interest here is to make it easier to integrate 3rd party tooling with Gogs.  I went to look at [screwdriver.cd](http://screwdriver.cd) (released by yahoo!) and started with their base and found that I failed immediately... well.. not quite immediately, but when I got to the 'get repo by (int) id'.  What's frustrating is that the methods on the Gogs side of things are all there... in fact... the [model-level method is there](https://github.com/gogits/gogs/blob/master/models/repo.go#L1469)... and it seems trivial to build the mapping... 

* [api routing update](https://github.com/patrickmcclory/gogs/commit/cbe7eb949e896d850fc778b3c435afcbe024a301)
* [repo update](https://github.com/patrickmcclory/gogs/commit/de052b8959083319163cbefa80ea3af4a7a9fdd9)

I've not tested this.. and forgive my golang familiarity.. but in terms of doing the right thing, I could get some legs out of this, but the further I dug in, the more detail I needed out of the base of the system... most of which wasn't available or filterable via the api.  

I'm going to start stubbing things out... and I'll put them here: https://github.com/patrickmcclory/gogs-api-draft

Happy to break it up and collaborate.

**Comment by patrickmcclory** [comment]
I have some models and a basic set of security model structures stubbed out--this is coming from the client's structs.. i'll work on modeling endpoints next. 

http://editor.swagger.io/#/?import=https://raw.githubusercontent.com/patrickmcclory/gogs-api-draft/master/spec.yaml

**Comment by brandoncole** [comment]
@patrickmcclory https://github.com/patrickmcclory/gogs-api-draft is very nice.  That it a fantastic start.

**Comment by patrickmcclory** [comment]
@kpande - is this something you're interested in continuing to pursue?
